src/mixin/TransientStorageArray.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.28;
+
+/// @title Transient Storage Array
+/// @author CoW DAO developers
+/// @notice Defines an interface to copy a (dynamic-size) bytes array from
+/// calldata to transient storage as well as to read it at a later point in the
+/// same transaction.
+/// @dev The array length is defined as a Solidity transient variable. The
+/// content of the array is stored word-by-word starting from the end of the
+/// transient storage space (slot `type(uint256).max`); each next word is stored
+/// in the transient storage slot immediately before. The endianness of
+/// each word has not been changed.
+/// The array is defined from the end to avoid conflicts with transient
+/// variables that could be relied on by the importing contract. If the contract
+/// that inherits this one only access transient storage through Solidity
+/// variables that are only manipulated through basic Solidity operations (i.e.,
+/// no assembly) then there should be no interferences with this contract.
+/// Otherwise, it's important to have a full understanding of the transient
+/// storage layout of the inheriting contract to avoid that the same transient
+/// slot is overridden when working with two unrelated variables.
+abstract contract TransientStorageArray {
+    uint256 constant BYTES_IN_WORD = 32;
+
+    /// @notice The number of bytes currently stored in the transient storage
+    /// array.
+    uint256 internal transient transientStorageLength;
+
+    function storeToTransientStorageArray(bytes calldata data) internal {
+        transientStorageLength = data.length;
+        uint256 storedBytes = 0;
+        while (storedBytes < data.length) {
+            // Each (transient) storage slot stores a full word.
+            bytes32 window;
+            // Note: if windowEnd > data.length, then the extra bytes of
+            // `window` are filled with arbitrary bytes (or zero bytes if after
+            // the calldata range). The variable `length` encodes the array
+            // boundaries, meaning that these extra bytes will be ignored when
+            // recovering the stored content.
+            assembly ("memory-safe") {
+                // `add` is not expected to overflow as sending a transaction
+                // with enough calldata to cause the offset to overflow would
+                // not fit the block gas limit.
+                window := calldataload(add(data.offset, storedBytes))
+            }
+
+            uint256 transientArraySlot;
+            unchecked {
+                // No overflow check: block gas limit avoids overflows.
+                transientArraySlot = type(uint256).max - (storedBytes / BYTES_IN_WORD);
```

Comment:
I guess since this is always starting from `uint256.max` calling `storeToTransientStorage()` twice would overwrite the original data. Would it make sense to instead implement this more like a bump allocator?
That way calling `storeToTransientStorage()` a second time would simply continue writing to transient storage and then return the bounds of the written array so that the caller knows where to read the storage from.

I supposed you could reserve the first byte of transient storage to store the currently stored number of bytes.